### DISCUSSION TITLE
Better way to map over zod object values

### DISCUSSION THREAD

**Discussion by tavoyne** [colinhacks/zod#1402] [Q&A]
Say I have an object that looks like that:

```tsx
const inputSchema = z.object({
  emailAddress: z.string().email(),
  password: z.string().min(5),
});
```

I'd like to create an _error_ object for this object. It'd have the same keys but all values set to `z.optional(z.array(z.string()))`.

If I was to create it manually, it'd look like that:
```tsx
const inputErrorSchema = z.object({
  emailAddress: z.optional(z.array(z.string())),
  password: z.optional(z.array(z.string())),
});
```

Here's what I came up with:
```tsx
const inputErrorSchema = z.object(
  Object.fromEntries(
    inputSchema.keyof().options.map((key) => {
      return [key, z.optional(z.array(z.string()))];
    })
  )
)
```

But it's verbose and looses type information.

Any better way?

**EDIT**

One way of not loosing type information:
```tsx
const inputErrorSchema = z.object(
  Object.fromEntries(
    inputSchema.keyof().options.map((key) => {
      return [key, z.optional(z.array(z.string()))];
    })
  ) as {
    [Property in ReturnType<
      typeof inputSchema.keyof
    >["options"][number]]: z.ZodOptional<z.ZodArray<z.ZodString, "many">>;
  }
)
```

But the level of verbosity just skyrocketed.

It'd be amazing to have a zod method to do that. It could be a one-liner: `inputSchema.map(z.optional(z.array(z.string())))`.

**Comment by nullndr** [comment]
Can't you just use [`.partial()`](https://github.com/colinhacks/zod#partial)?

**Reply by tavoyne** [reply]
`.partial()` is just for marking some (or all) properties as optional.

That's not what I'm trying to achieve. See the example in my question:
>  If I was to create it manually, it'd look like that:
>
>  ```ts
>  const inputErrorSchema = z.object({
>    emailAddress: z.optional(z.array(z.string())),
>    password: z.optional(z.array(z.string())),
>  });
>  ```

**Reply by nullndr** [reply]
You are making the properties optionals

**Reply by tavoyne** [reply]
Yeah but I'm not _just_ doing that. I'm also making them an array of strings.

**Reply by nullndr** [reply]
Oh sorry, I didn't see it

**Reply by nullndr** [reply]
I don't think zod currently has something like that, the easiest way for me is doing it manually

**Comment by jacknewberry** [comment]
I also have a use for this object mapping functionality. @tavoyne I will use your workaround.
